### ISSUE TITLE
Molarity Alert Manager

### ISSUE THREAD

**Issue by twant** [phetsims/molarity#106]
@zepumph per our conversation yesterday, I created an alert manager that only manages the creation of utterances and adding them to the utteranceQueue. If you have time, would love your thoughts -- I think it's cleaner, but I'm not sure if the alert manager should be doing more work.

**Comment by zepumph** [comment]
Review of the alert manager refactor plus anything I saw along the way:

* https://github.com/phetsims/molarity/commit/5a6cde101733e875178094e967bdf7fa9a7189c8 is a very large commit, and it is sorta hard to see what was going on in it by just looking at the diff. When it is possible to commit in more modular, and specific ways, please do. I know, especially during refactoring times like these, how it can be hard to manage that, and a changeset can get away from you. I have definitely committed that large of a commit, and will do it again! But just something to think about.
* (I think we discussed this before but just in case) `getSoluteAmountChangedValueText` and others like it are very confusingly named because they do everything a getter does, but then also alert on the side. Please rename to be clear about both functions of the method.
* I see that a single Utterance is being used for both sliders. This is likely a design decision, but I could imagine a weird case (at least I think its weird) when switching between sliders quickly where you don't get an alert update from one because changing the other overwrites it. Maybe we should have two utterances, one for each.
* (read this and the point below before doing one of them) I really like how you factored things out into the alert manager. Since Molarity is just a single screen sim, and the alert manager doesn't have dependencies that need to be passed to it in some startup order, I think we can just make it a singleton. Which would look like this at the end of the file:
`return molarity.register( 'molarityAlertManager', new MolarityAlertManager() );`
The reason I'm ok with making it a singleton when we just changed all the other globals back to instances passed around is because this looks much more like the traditional PhET singleton, and the code could be cleaned up by not needing to pass it around to all the describers that need it. 
* (read this and the point above before doing one of them) Instead of making the alert manager a  singleton, you could factor out more responsibility to the alert manager. For example, notice how GravityForceLabAlertManager is passed describers, and the methods to "alert" something don't take arguments, or at least don't take the alert as an argument. And so then you can simplify the call site because the alert manager takes care of formulating the strings (by calling the describer methods itself). I think I may prefer this approach generally, but I know your aria-valuetext and alert text creation may be pretty closely tied, so if you feel like you can't factor this much logic into the alert manager, then the above review bullet is good too.

**Comment by twant** [comment]
Thanks for this review @zepumph! Just pushed a few commits to address these. I gave MolarityAlertManager responsibility for the methods in ConcentrationDescriber, which does feel cleaner. See what you think -- I think a singleton could work well as well. 

The only one of your points I haven't implemented yet is the separate of slider utterances -- they're pretty intertwined right now in the implementation, so I'll put some thought into the cleanest way to separate them!

**Comment by zepumph** [comment]
> I think a singleton could work well as well.

I think I prefer your commits to a singleton, and https://github.com/phetsims/molarity/commit/47b35f5cf8e4fa5c7b5b351b5af709b67c48ad34#diff-01db8c6093f1395a996b21d075aa4d87R34 wouldn't work with a singleton, because it needs to have model state passed to it.


* A small thing. In `MolarityAlertManager.alertSliderQualitative()`, the string `qualitativeValueTextPatternString` takes an optional `stateInfo`, but the tail end of the string literal includes a space at the end. Perhaps if the `{{stateInfo}}` var filled in that space instead of having it dangling, that would be a bit cleaner.
* The dependency hierarchy is a bit strange to me, although I understand why it is the way it is. The alert manager needs the concentration describer, and is called by a describer. To call the alert manager from the describer breaks the logical division I have created in my mind between to two until this point, not that that is necessarily bad. Let's rethink this a bit. Take `SoluteAmountDescriber`, I see many methods in this type that do not call `StringUtils.fillin`. What if getSoluteAmountDescriptionsAndAlert` and `getQualitativeSoluteAmountDescriptionsAndAlert`, getQuantitativeSoluteAmountDescriptionsAndAlert` were all inlined in the screen view, or in another place (not a describer). Here is patch exploring that briefly. At first glance it isn't prohibitively complicated to not just be straight in the screen view, though I haven't looked to closely.

    <details>
    
     ```
    Index: js/molarity/view/MolarityScreenView.js
    IDEA additional info:
    Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
    <+>UTF-8
    ===================================================================
    --- js/molarity/view/MolarityScreenView.js	(revision d893cfdbebd28056e560507adca4b6a178bd1b09)
    +++ js/molarity/view/MolarityScreenView.js	(date 1559591127884)
    @@ -143,7 +143,36 @@
           soluteAmountAccessibleNameString,
           soluteAmountHelpTextString,
           () => soluteAmountDescriber.getOnFocusSoluteAmountValueText(),
    -      () => soluteAmountDescriber.getSoluteAmountDescriptionsAndAlert(),
    +      () => {
    +        if ( concentrationDescriber.isNewSaturationState() ) {
    +          this.alertManager.alertSaturation( concentrationDescriber.getSaturationChangedString() );
    +        }
    +        if ( useQuantitativeDescriptionsProperty.value ) {
    +
    +          // alerts
    +          this.alertManager.alertSliderQuantitative( this.isInitialSoluteAmountAlert );
    +          if ( this.isInitialSoluteAmountAlert ) {
    +            this.isInitialSoluteAmountAlert = false;
    +          }
    +
    +          // aria-valueText
    +          return this.concentrationDescriber.getQuantitativeValueText( this.isInitialSoluteAmountAlert,
    +            this.getCurrentSoluteAmount() );
    +
    +        }
    +        else {
    +          // alerts (a special alert is read out when there is no solute in the beaker)
    +          if ( this.solution.soluteAmountProperty.value <= 0.001 ) {
    +            this.alertManager.alertSlider( noSoluteAlertString );
    +          }
    +          else {
    +            this.alertManager.alertSliderQualitative( this.getSoluteAmountChangeString(), this.soluteAmountRegionChanged );
    +          }
    +
    +          // aria-valueText
    +          return this.getSoluteAmountState();
    +        }
    +      },
           () => soluteAmountDescriber.setInitialSoluteAmountAlert(),
           model.solution.soluteProperty
         );
    Index: js/molarity/view/describers/SoluteAmountDescriber.js
    IDEA additional info:
    Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
    <+>UTF-8
    ===================================================================
    --- js/molarity/view/describers/SoluteAmountDescriber.js	(revision d893cfdbebd28056e560507adca4b6a178bd1b09)
    +++ js/molarity/view/describers/SoluteAmountDescriber.js	(date 1559591127890)
    @@ -181,12 +181,7 @@
          * @returns {null|string}
          */
         getSoluteAmountDescriptionsAndAlert() {
    -      if ( this.concentrationDescriber.isNewSaturationState() ) {
    -        this.alertManager.alertSaturation( this.concentrationDescriber.getSaturationChangedString() );
    -      }
    -      return this.useQuantitativeDescriptionsProperty.value ?
    -             this.getQuantitativeSoluteAmountDescriptionsAndAlert() :
    -             this.getQualitativeSoluteAmountDescriptionsAndAlert();
    +
         }
     
         /**
    @@ -216,16 +211,7 @@
          */
         getQualitativeSoluteAmountDescriptionsAndAlert() {
     
    -      // alerts (a special alert is read out when there is no solute in the beaker)
    -      if ( this.solution.soluteAmountProperty.value <= 0.001 ) {
    -        this.alertManager.alertSlider( noSoluteAlertString );
    -      }
    -      else {
    -        this.alertManager.alertSliderQualitative( this.getSoluteAmountChangeString(), this.soluteAmountRegionChanged );
    -      }
     
    -      // aria-valueText
    -      return this.getSoluteAmountState();
         }
       }
     
    ```
    <details>

>The only one of your points I haven't implemented yet is the separate of slider utterances -- they're pretty intertwined right now in the implementation, so I'll put some thought into the cleanest way to separate them!

* very similar to the above bullet and is the fact that you haven't separated out the logic of alerts from the aria-valuetext, but it was only after exploring the above patch that I realized that from this intertwined logic you are actually powering alerts whenever aria-valuetext is updated for the AccessibleValueHandler (in this case Slider). This is very atypical and quite dangerous. aria-valuetext is an on demand, description, and so it was designed to potentially be updated a number of times when the slider isn't focused such that it is kept in sync with the model. Under the current logic, that would result in tons of alerts, not just when the original model Property changed. Instead of putting the alerting logic in the aria value text creator function, it should be in a direct listener on the model Property you care about. Something like: `model.solution.soluteAmountProperty.link( ()=> doTheAlerting()`. We can talk more about the specifics about why entangling these two is so buggy, but from an implementation standpoint it largely has to do with the newly created AccessibleValueHandler option `a11yDependencies`.
* There is a line in the screen view like:
    ```js
    // Determines whether qualitative or quantitative a11y descriptions are used.
    const useQuantitativeDescriptionsProperty = new DerivedProperty( [ valuesVisibleProperty ], () => {
      return valuesVisibleProperty.value;
    } );
    ```
    This can be simplified because the function returns values, like so:
    ```js
    const useQuantitativeDescriptionsProperty = new DerivedProperty( [ valuesVisibleProperty ],  
      currentValuesVisible => currentValuesVisible );`
    ```
    Or you can use a lodash function like `_.identity` that will return the first arg. https://lodash.com/docs/4.17.11#identity

**Comment by zepumph** [comment]
Unassigning for now. Let me know if I can help more.

**Comment by twant** [comment]
I think these review comments have been addressed, and we've been over the alert manager several times since then. I'm going to close this for now.
